Snapshot store

AlexandruCihodaru opened the pull request:
This PR is a first step in an attempt to fix https://github.com/qdrant/qdrant/issues/3324. 
To be able to add S3 support and other storage back ends later I defined a trait that will contain all needed methods to handle created snapshots. I took a simplifying assumption that the snapshots will first be created in a temp local place and after the snapshotting is completed we will move the snapshot and checksum.
After S3 support is added, it should be straight forward to implement support for additional storage back ends, the only needed action being implementing the trait for the new type. 

To keep the PR small and easy to review, currently it only handles create_snapshot,delete,list (collection/full/shard). If the direction is good others could be implemented in subsequent PRs without breaking compatibility.

### All Submissions:

* [x] Contributions should target the `dev` branch. Did you create your branch from `dev`?
* [x] Have you followed the guidelines in our Contributing document?
* [x] Have you checked to ensure there aren't other open [Pull Requests](../../../pulls) for the same update/change?

<!-- You can erase any parts of this template not applicable to your Pull Request. -->

### New Feature Submissions:

1. [x] Does your submission pass tests?
2. [x] Have you formatted your code locally using `cargo +nightly fmt --all` command prior to submission?
3. [x] Have you checked your code using `cargo clippy --all --all-features` command?

### Changes to Core Features:

* [x] Have you added an explanation of what your changes do and why you'd like us to include them?
* [ ] Have you written new tests for your core changes, as applicable?
* [x] Have you successfully ran tests with your changes locally?

AlexandruCihodaru commented on lib/storage/src/types.rs:
```diff
@@ -53,7 +53,7 @@ pub struct StorageConfig {
     #[serde(default = "default_snapshots_path")]
     #[validate(length(min = 1))]
-    pub snapshots_path: String,
+    pub snapshots_path: Option<String>,
```
This change is needed because in the future there could be either snapshot_path for local storage either some details for the storage back end. It would have been cleaner with an enum in here but I believe it would break backwards compatibility.

generall replied:
:check:

AlexandruCihodaru replied:
@generall an enum would mean to also have a custom deserialization implementation and everything, overall I felt that it was cleaner how I implemented it here, but for reference this is how it would look like https://github.com/AlexandruCihodaru/qdrant/commit/887d42631996c554e419f8b54eb2b95c397914b6

AlexandruCihodaru commented on lib/collection/src/collection/mod.rs:
```diff
@@ -141,6 +143,7 @@ impl Collection {
             this_peer_id,
             path: path.to_owned(),
             snapshots_path: snapshots_path.to_owned(),
+            snapshot_storage_manager: Self::create_snapshot_storage_manager(snapshots_path),
             channel_service,
```
I think that in the future snapshots_path could be remove so we will avoid info duplication, the path can be retrieved from the snapshot_storage_manager.

AlexandruCihodaru commented on lib/collection/src/collection/mod.rs:
```diff
@@ -50,6 +51,7 @@ pub struct Collection {
     this_peer_id: PeerId,
     path: PathBuf,
     snapshots_path: PathBuf,
+    snapshot_storage_manager: Box<dyn SnapshotStorage>,
```
This could be static dispatch by making Collection generic over SnapshotStorage. We could switch to it if snapshotting happens often because it is cheaper.

generall commented on lib/collection/src/collection/snapshots.rs:
```diff
@@ -135,23 +134,10 @@ impl Collection {
-        snapshot_ops::get_snapshot_description(&snapshot_path).await
+        let snapshot_storage_manager = &self.snapshot_storage_manager;
+        snapshot_storage_manager
+            .store_file(&snapshot_path_tmp_move, snapshot_path.as_path())
```
Actual moving to another location (mount or S3) happens n the `fs::copy` statement above. 
Therefore it should also be habdled by `snapshot_storage_manager`

generall replied:
Not related to this PR, but checksum should be computed before copying

AlexandruCihodaru replied:
Right. I moved it, also moved the checksum computation before copying.

generall commented on lib/collection/src/shards/transfer/snapshot.rs:
```diff
@@ -188,10 +189,17 @@ pub(super) async fn transfer_snapshot(
         "Local shard must be a queue proxy"
     );
 
+    let snapshot_storage_manager = Collection::create_snapshot_storage_manager(snapshots_path);
```
The fact that we are using `create_snapshot_storage_manager` from shards suggesrts that it should not be part of `Collection`.
Ideally shards should not know anything about Collection existance, otherwise it is circular dependency.

AlexandruCihodaru replied:
Do you think that lib/common/common would be a good location?

generall replied:
I would do it as a `::new` with config and path as arguments. Since we do have access to both in all cases

AlexandruCihodaru replied:
I think we don't have access to StorageConfig (which will hold the info for S3/LocaFS) from the shard holder, it does have SharedStorageConfig, so we could update that structure I think :shrug:

generall replied:
`transferring_shard` have access to `shared_storage_config`. We can have a method to create SnapshotManager in both Collection and ShardReplicaSet

AlexandruCihodaru replied:
Yes but SharedStorageConfig https://github.com/qdrant/qdrant/blob/d0baa5ef4a4bc2c597b28d0e1f1011262be8ed4d/lib/collection/src/operations/shared_storage_config.rs#L17 does not have the needed info for handling snapshot transfer

generall replied:
Good point. I would add this info into `SharedStorageConfig` then

AlexandruCihodaru commented on lib/storage/src/content_manager/snapshots/mod.rs:
```diff
@@ -44,9 +41,15 @@ pub async fn do_delete_full_snapshot(
+        .clone()
+        .toc()
+        .storage_config
+        .to_shared_storage_config(false);
```
Here I am creating the shared storage config from the storage config to avoid duplicating the *snapshot_manager* method.

AlexandruCihodaru commented:
Hey @generall, I have addressed your comments. In a subsequent PR we should also move the get_snapshot_description and get_snapshot/checksum_path into the trait because we will have different implementations for the LocalFS and S3.

AlexandruCihodaru commented on lib/collection/src/collection/mod.rs:
```diff
@@ -46,7 +46,7 @@ pub struct Collection {
     pub(crate) shards_holder: Arc<LockedShardHolder>,
     pub(crate) collection_config: Arc<RwLock<CollectionConfig>>,
-    pub(crate) shared_storage_config: Arc<SharedStorageConfig>,
+    pub shared_storage_config: Arc<SharedStorageConfig>,
```
Definitely.

AlexandruCihodaru commented:
Hey @generall  I was wondering if the desired outcome is to use S3 like a "persistent" storage or to use it for any snapshot related actions. If we targeting the first option than I think it is possible to bring locally from S3 the snapshot dir before initializing the TableOfContent in here https://github.com/qdrant/qdrant/blob/d0baa5ef4a4bc2c597b28d0e1f1011262be8ed4d/lib/storage/src/content_manager/toc/mod.rs#L94 and I believe that there will not be a lot of additional changes after this PR.

generall commented:
/tip $200
